File: 32-LongestValidParentheses/32-LongestValidParentheses.py
# Last updated: 8/1/2025, 10:42:07 PM
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        st = []
        max_len = 0
        st.append(-1)  # Acts as a base index

        for i in range(len(s)):
            if s[i] == '(':
                st.append(i)
            else:
                st.pop()
                if not st:
                    st.append(i)
                else:
                    max_len = max(max_len, i - st[-1])
        
        return max_len

File: 32-LongestValidParentheses/32-LongestValidParentheses.java
// Last updated: 8/1/2025, 10:42:14 PM
class Solution {
    public int longestValidParentheses(String s) {
      Stack<Integer> st=new Stack();
      int len=0;
        st.push(-1);
      for(int i=0;i<s.length();i++){
        if(s.charAt(i)=='('){
            st.push(i);
        }else{
            st.pop();
            if(st.isEmpty()){
                st.push(i);
            }else{
                len=Math.max(len,i-st.peek());
            }
        }
      }
        return len;
    }
    }